src/main/java/com/google/devtools/build/lib/runtime/BlazeOptionHandler.java

```diff
@@ -257,11 +265,88 @@ private void parseArgsAndConfigs(List<String> args, ExtendedEventHandler eventHa
     expandConfigOptions(eventHandler, commandToRcArgs);
   }
 
+  /**
+   * {@link ExtendedEventHandler} override that passes through "normal" events but not events that
+   * would go to the build event proto.
+   *
+   * <p>Starlark flags are conceptually options but still need target pattern evaluation. If we pass
+   * {@link #post}able events from that evaluation, that would produce "target loaded" and "target
+   * configured" events in the build event proto output that consumers can confuse with actual
+   * targets requested by the build.
+   *
+   * <p>This is important because downstream services (like a continuous integration tool or build
+   * results dashboard) read these messages to reconcile which requested targets were built. If they
+   * determine Blaze tried to build {@code //foo //bar} then see a "target configured" message for
+   * some other target {@code //my_starlark_flag}, they might show misleading messages like "Built 3
+   * of 2 requested targets.".
+   *
+   * <p>Hence this class. By dropping those events, we restrict all info and error reporting logic
+   * to the options parsing pipeline.
+   */
+  private static class NonPostingEventHandler implements ExtendedEventHandler {
+    private final ExtendedEventHandler delegate;
+
+    NonPostingEventHandler(ExtendedEventHandler delegate) {
+      this.delegate = delegate;
+    }
+
+    @Override
+    public void handle(Event e) {
+      delegate.handle(e);
+    }
+
+    @Override
+    public void post(ExtendedEventHandler.Postable e) {}
+  }
+
+  /**
+   * Lets {@link StarlarkOptionsParser} convert flag names to {@link Target}s through {@link
+   * TargetPatternPhaseValue}.
+   *
+   * <p>This is used for top-level flag parsing, outside any {@link SkyFunction}.
+   */
+  public static class SkyframeExecutorTargetLoader
```

Comment:
Do we need a new inner class? Can the BlazeOptionHandler implement BuildSettingLoader directly, or just define `loadBuildSetting` in BlazeOptionHandler and pass in a method reference?

Reply:
`CanonicalizeCommand` also references this, which is why I made it its own entity: https://github.com/bazelbuild/bazel/blob/f2ccd3c1362546b8d346ed19a1c6eff3951dfce9/src/main/java/com/google/devtools/build/lib/runtime/commands/CanonicalizeCommand.java#L201